pypeit/core/flexure.py

```diff
@@ -55,68 +59,223 @@ def spat_flexure_shift(sciimg, slits, debug=False, maxlag = 20):
+                  '  Assuming there is NO SPATIAL FLEXURE.'+msgs.newline() + 'If a flexure is expected, '
+                  'consider either changing the maximum lag for the cross-correlation, '
+                  'or the "spat_flexure_sigdetect" parameter, or use the manual flexure correction.')
+
         return 0.
-    
-    # Find the peak
-    xcorr_max = np.interp(pix_max, np.arange(lags.shape[0]), xcorr_norm)
-    lag_max = np.interp(pix_max, np.arange(lags.shape[0]), lags)
-    msgs.info('Spatial flexure measured: {}'.format(lag_max[0]))
+    lag0 = np.where(lags == 0)[0][0]
+    shift = round(pix_max[0] - lag0, 3)
+    msgs.info('Spatial flexure measured: {}'.format(shift))
 
     if debug:
-        plt.figure(figsize=(14, 6))
-        plt.plot(lags, xcorr_norm, color='black', drawstyle='steps-mid', lw=3, label='x-corr')
-        plt.plot(lag_max[0], xcorr_max[0], 'g+', markersize=6.0, label='peak')
-        plt.title('Best shift = {:5.3f}'.format(lag_max[0]) + ',  corr_max = {:5.3f}'.format(xcorr_max[0]))
+        # 1D plot
+        xvals = np.arange(corr_slits.size)
+        plt.figure(figsize=(9, 8))
+        scale = np.nanmax(corr_slits) / np.nanmax(corr_sci)
+        plt.suptitle(f'Shift={shift:.1f} pixels', fontsize=18)
+        plt.plot(xvals, corr_slits, 'k', label='slitmask edge peaks')
+        plt.plot(xvals, np.roll(scale*corr_sci, -int(shift)), '--r', label='shifted science edge peaks')
         plt.legend()
+        plt.tight_layout()
         plt.show()
 
-    #tslits_shift = trace_slits.shift_slits(tslits_dict, lag_max)
-    # Now translate the tilts
+        # 2D plot
+        spat_flexure_qa(sciimg, slits, shift, gpm=np.logical_not(bpm), vrange=qa_vrange)
+
+    if qa_outfile is not None:
+        # Generate the QA plot
+        msgs.info("Generating QA plot for spatial flexure")
+        spat_flexure_qa(sciimg, slits, shift, gpm=np.logical_not(bpm), vrange=qa_vrange, outfile=qa_outfile)
+
+    return shift
+
 
-    #slitmask_shift = pixels.tslits2mask(tslits_shift)
-    #slitmask_shift = slits.slit_img(flexure=lag_max[0])
+def spat_flexure_qa(img, slits, shift, gpm=None, vrange=None, outfile=None):
+    """
+    Generate QA for the spatial flexure
+
+    Args:
+        img (`numpy.ndarray`_):
+            Image of the detector
+        slits (:class:`pypeit.slittrace.SlitTraceSet`):
+            Slits object
+        shift (:obj:`float`):
+            Shift in pixels
+        gpm (`numpy.ndarray`_, optional):
+            Good pixel mask (True = Bad)
+        vrange (:obj:`tuple`, optional):
+            Tuple with the min and max values for the imshow plot
+        outfile (:obj:`str`, optional):
+            Path to the output file where the QA is saved.  If None, the QA is
+            shown on screen and not saved.
 
+
+    """
+    debug = True if outfile is None else False
+    # TODO: should we use initial or tweaked slits in this plot?
```

Comment:
I think it's the tweaked slits that you're after. As you noted above on line 85, whatever is used here needs to be consistent with that choice. I think once the tweaked slits are available, they get used.

Reply:
yes, I believe that is correct.

Reply:
Well in the code above I'm still using the initial slits (not the tweaked ones) since:
`slitmask = slits.slit_img(initial=True, exclude_flag=slits.bitmask.exclude_for_flexure)`
It seems to me that the initial ones should work better in the cross-correlation, and therefore this plot should be consistent with that. But I do not want users to get confused when seeing this plot and comparing with what is actually use for reduction, i.e., tweaked slits if `tweak_slits=True`